### ISSUE TITLE
Rework GameCube controller support

### ISSUE THREAD

**Issue by meeq** [DragonMinded/libdragon#218]
*Extracted from https://github.com/DragonMinded/libdragon/pull/215#issuecomment-955559468*

While reviewing the controller subsystem routines for GameCube controllers to update the Doxygen comments, I have collected some notes concerning the usefulness and ergonomics of the current implementation:

1. Players that want to use a GameCube controller on N64 will need to make a passive GCN->N64 controller adapter for each controller in order to use it.
    * GameCube controllers do not identify as N64 controllers, use separate Joybus commands, and have a completely different data structure, so commercially-available GCN->N64 adapters actively convert the Joybus commands for compatibility with existing Nintendo 64 software.
    *  The use-case for using GameCube Joybus commands on a Nintendo 64 is extremely esoteric and was never supported by Nintendo or any third-party.
2. `controller_read_gc` and `controller_read_gc_origin` are not integrated with the rest of the controller subsystem.
    * A developer would need completely separate controller-handling code-paths to support GameCube controllers.
3. `struct controller_data` is wasteful, confusing, and unnecessary:
    * You will never get N64 and GCN controller data in the same result, so half of the struct is always padding.

I don't necessarily think we should remove the GC controller commands completely, but I think they need to be removed from the rest of the controller subsystem and flagged as highly-experimental if it even makes sense to keep it at all.

I strongly doubt that any developer except the original author is actually using these GC controller functions, so it might make sense to deprecate them and rework the developer experience (or just remove them entirely). Right now the usage is a lot of pain for very little gain.

**Comment by meeq** [comment]
Linking this with #48 where GameCube controller support was introduced

**Comment by meeq** [comment]
https://github.com/clbr/n64gctest is using the GameCube controller support functions

**Comment by meeq** [comment]
Pinging @clbr to get some more context on the intentions of this functionality and what the impact of deprecation/rework would be for their projects.

**Comment by meeq** [comment]
Linking to the n64-dev Discord discussion about GC controllers on N64: https://discord.com/channels/205520502922543113/205520502922543113/903630132294611044

**Comment by clbr** [comment]
I cannot see that discord discussion, please post it on pastebin or here. Further, I'm appalled at your approach in general; "this homebrew is doing new, exciting things that no released software did - therefore we must remove it". Seriously, WTF?

1. The use case of GC controllers on N64 is extremely useful, and extremely positive PR for any title *exactly* because it was never supported by Nintendo. The compatibility of the two consoles' controller buses must be used to do interesting things.
2. Duh. The GC controllers have a different button layout - of course it requires different code paths! You cannot expect one stick and two sticks to share the same moving code.
3. Separate the structs if you feel like it.

"Right now the usage is a lot of pain for very little gain." It is very easy and provides something excellent that Nintendo itself never provided.

> what the impact of deprecation/rework would be for their projects.
This gets political, but the impact would be zero for two reasons.
- the author of libdragon started accepting poor quality PRs at one point, severely reducing the library's quality and forcing me to review each commit manually
- the author of libdragon took a political stance I do not agree with (SJW naming of the branches); I want nothing to do with any project that does this

For these two reasons, I am already using a private fork, and will continue to do so. So if you want to remove this functionality because no current libdragon user makes use of it - despite the obvious pluses to allowing projects to use a better controller with two sticks - go ahead.

**Comment by DragonMinded** [comment]
Hey man, fuck off. Go to your fork, don't come back here. You said it yourself, you don't want anything to do with any project like this. So keep your fork, stay away from here, do not comment and do not stir shit up anymore. We don't need people like you that make a huge fuss about a branch rename to stay inclusive, nor do we need people who want to take it upon themselves to trash other user contributions because it doesn't hold up to their personal standards. Go away.

**Comment by clbr** [comment]
I was asked to comment, and I kept my comment respectful. Unlike you.

**Comment by meeq** [comment]
Closing this. Sorry, I did not realize I was opening a can of worms here. I just want to make cool stuff for the Nintendo 64.

I think GameCube controller support in LibDragon is an interesting feature, and I want to help make it more accessible. To that end, I think the best path forward for GC support would be:

* Document the required passive GC->N64 controller adapter. It's really simple, but it's not documented.
* Move the raw GC controller code out of the controller subsystem into a separate `gc_controller` module.
* Attempt to integrate GameCube controllers with the LibDragon controller subsystem so that everyone gets basic GC->N64 controller support "for free".

There's a chicken-and-egg problem here with the adapters: no one has an adapter because no games support it, and no games support it because no one has the adapter. My hope is that if GC controller support in LibDragon is better-documented and "just works", then more people will make/sell these passive adapters.

The concession with "for free"/"just works" support is that there are more inputs on a GC controller than N64, so mapping GC data to the N64 controller struct means losing X and Y and crushing the C-stick down to 4 buttons. Developers who want to go the extra mile could still detect the GC controller and use the raw commands, just like how things work currently.
